Ipf trig

xznhj8129 opened the pull request:
### **User description**
Add support for arc function IPF LCs (see https://github.com/iNavFlight/inav/pull/11179)


___

### **PR Type**
Enhancement, New Feature


___

### **Description**
- Add support for inverse trigonometric functions (acos, asin, atan2)

- Replace hardcoded operation numbers with named constants for maintainability

- Update error messages and diagnostics to reflect new supported functions

- Define new logic operators 57, 58, 59 for arc trigonometry operations


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["New Arc Trig Functions"] -->|acos, asin, atan2| B["Expression Generator"]
  B -->|Generate Logic Commands| C["Logic Operators 57-59"]
  C -->|Map to Operations| D["INAV Constants"]
  A -->|Update Type Definitions| E["TypeScript Declarations"]
  B -->|Update Diagnostics| F["Editor Validation"]
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>helpers.js</strong><dd><code>Add arc trig functions and refactor operation constants</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-1faf75d1ec431c7b24fb58463320f6e3cd35962d70ddb9ebde7fa94fcc65556e">+42/-14</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>expression_generator.js</strong><dd><code>Implement code generation for arc trigonometric functions</code></dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-6ae761b8a43095786c28c6d9bfe73f42ec09ab2c6d3a5d45095e1673f07a7ca0">+26/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>logicConditionOperators.js</strong><dd><code>Define new logic operators for arc trigonometry</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-bc560b8c8960c536beec20d7ecde59f3e217ec6c2f7cf35d7c7234bcf7dfd794">+19/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>diagnostics.js</strong><dd><code>Update diagnostics to support arc trigonometric functions</code></dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-33b7e2ed92cc9eaa0cc9a5fa1de732b5069a282ff4e1dffe4ca96cf93719e524">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>inav_constants.js</strong><dd><code>Add operation constants for arc trigonometric functions</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-a9ad7650543b4ee3ff4c9f4eb1baa452c43ff492e8d62034a115c71f40d2a95a">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>types.js</strong><dd><code>Add TypeScript type declarations for arc functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-e7c92fcb37e2683b50013001b5677e7167c18fdc19e4519bc711b752da696416">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>inav_SITL</strong></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-31602c497002da343849ba5df2668d5ceadb62a18408330256d83a60eb1fd923">[link]</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>inav_SITL</strong></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-fcfcb5d872a9287876c8d61ad25e4e6ca9a1209e6209c81fcdf0427affd831de">[link]</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>inav_SITL</strong></td>
  <td><a href="https://github.com/iNavFlight/inav-configurator/pull/2478/files#diff-911063c252df7082d2bf2292955c54efcf4d52ea73f750198a95e1e08225f9ff">[link]</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

xznhj8129 committed:
change for ipf trig

sensei-hacker commented:
It looks like you did a lot of great work here. Thanks.
Adding to the Qodo comments comments above, Claude suggests maybe checking the following.
Of course, Claude can be wrong, or you can choose to do things differently.  We just want to make sure any choices are intentional.

```

  CRITICAL — Block merge
                                                                                                                                       
  1. Wrong opcode numbers (inav_constants.js, logicConditionOperators.js)                                                              
  Firmware PR #11179 already took opcode 57 for SET_ALTITUDE_TARGET. This PR assigns ACOS=57, ASIN=58, ATAN2=59 — but the correct      
  values are ACOS=58, ASIN=59, ATAN2=60. Sending wrong opcodes silently commands the wrong Logic Condition operation on the FC.        
                      
  2. Broken import path (helpers.js line 12)                                                                                           
  // Wrong — resolves to a non-existent path:
  import { OPERATION } from '../../transpiler/transpiler/inav_constants.js';                                                           
                                                                            
  // Correct:                                                                                                                          
  import { OPERATION } from '../../transpiler/inav_constants.js';
  This will crash the transpiler module on load.                                                                                       
                      
  ---                                                                                                                                  
  IMPORTANT — Should fix
                                                                                                                                       
  3. acos/asin input scaling is undocumented (expression_generator.js)
  The firmware divides operandA by operandB, and when operandB=0 it defaults to 1000. So Math.acos(0.5) actually computes              
  acos(0.5/1000) ≈ acos(0) = 90°, not acos(0.5) = 60°. The API docs say params: ['ratio'] implying -1..1 input, but the real expected  
  range is -1000..1000. Needs either a documentation fix or explicit operandB=1000.                                                    
                                                                                                                                       
  4. abs inavOperation silently removed (helpers.js)                                                                                   
  All sibling math functions got their hardcoded numbers replaced with named constants — correct. But abs had the inavOperation field
  removed entirely with no comment explaining why (it's because abs is emulated via MAX(x, SUB(0,x)) since firmware opcode 32 is       
  actually SET_OSD_LAYOUT). Should have an inline comment.
                                                                                                                                       
  5. SET_ALTITUDE_TARGET (57) missing from OPERATION_NAMES (inav_constants.js)                                                         
  Fixing the opcode numbers (item 1) will also require adding this missing entry.

```

xznhj8129 committed:
Merge upstream/maintenance-10.x into ipf_trig; align arc-trig LC numbers

Resolve the inav_constants.js OPERATION_NAMES conflict and bring the
branch up to date with maintenance-10.x (it was 517 commits behind).

The firmware that actually merged (iNavFlight/inav#11179, plus a
SET_ALTITUDE_TARGET=57 inserted ahead of it) assigns the arc-trig
logic conditions as ACOS=58, ASIN=59, ATAN2=60 — one higher than the
57/58/59 this PR originally used. Renumber ACOS/ASIN/ATAN2 in the
OPERATION constants, OPERATION_NAMES, and LOGIC_OPERATORS to match,
so generated logic commands map to the correct firmware operations.

xznhj8129 committed:
Add SET_ALTITUDE_TARGET (LC operation 57)

Firmware maintenance-10.x defines LOGIC_CONDITION_SET_ALTITUDE_TARGET=57
(navigationSetAltitudeTargetWithDatum): operand A = altitude datum flag
(0 takeoff / 1 MSL / 2 terrain), operand B = target altitude. It was
missing from the configurator, leaving a gap at 57.

Register it in the OPERATION constants, OPERATION_NAMES, and the
LOGIC_OPERATORS table (two operands, boolean output) so it is selectable
and labeled in the Logic Conditions editor.

xznhj8129 committed:
fix + set_alt

xznhj8129 committed:
reorder
